### PR TITLE
Update interfaces with a lambda style approach

### PR DESCRIPTION
**Pull request by shadowhand**
Based on feedback from the community, a lambda style approach offers
better runtime checking and provides a more resilient foundation for
middleware containers and components.

Also clarifies the goals in the meta document accordingly.

http://blog.ircmaxell.com/2016/05/all-about-middleware.html
https://groups.google.com/d/msg/php-fig/sYecuqdYIkY/Gh5ATLn_EAAJ

### CODE REVIEW

**Comment by shadowhand** [comment]
Please post all feedback on this mailing list thread: https://groups.google.com/d/msg/php-fig/sYecuqdYIkY/Gh5ATLn_EAAJ

File: proposed/http-middleware/middleware.md [hunk 1]
```diff
@@ -2,102 +2,203 @@ HTTP Middleware
...
 
 ## 1. Specification
 
 An HTTP middleware component is an individual component participating, together
 with other middleware components, in the processing of an incoming HTTP Request
 and the creation of a resulting HTTP Response, as defined by PSR-7.
 
-Middleware components MUST implement this interface.
+Middleware consumers such as frameworks and middleware stack containers MUST
+type hint against one of the middleware interfaces:
```
Comment (kelunik) [inline proposed/http-middleware/middleware.md]: _type hint_ should be _type declaration_. These are not hints, they're enforced. `/** @var Type */ is a hint, but these are declarations. See also http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration.

File: proposed/http-middleware/middleware.md [hunk 2]
```diff
@@ -2,102 +2,203 @@ HTTP Middleware
...
-{
-    // ...
-}
-```
+Generally consumers SHOULD type hint against `MiddlewareInterface` unless the
+consumer only processes client requests.
+
+### 1.1 Containers
+
+And HTTP middleware container is an object that holds multiple middleware
```
Comment (kelunik) [inline proposed/http-middleware/middleware.md]: And → An

**Comment by pmjones** [comment]
@shadowhand Happy to merge whenever you feel it's ready.

**Comment by shadowhand** [comment]
@pmjones I have fixed the typos that @kelunik pointed out, you can merge.

**Comment by kelunik** [comment]
@shadowhand _type hint_ is used multiple times, should always be _type declaration_.

**Comment by shadowhand** [comment]
@kelunik ah, my mistake. I will make another PR to correct that.

**Pull request merged**